drivers/usb/device/usb_dc_stm32.c

```diff
@@ -428,7 +428,11 @@ static int usb_dc_stm32_init(void)
 	usb_dc_stm32_state.pcd.Init.phy_itface = PCD_PHY_EMBEDDED;
 #endif
 	usb_dc_stm32_state.pcd.Init.ep0_mps = USB_OTG_MAX_EP0_SIZE;
+#if CONFIG_USB_VBUS_SENSING_STM32
+	usb_dc_stm32_state.pcd.Init.vbus_sensing_enable = ENABLE;
```

Comment:
Does it enable HAL_PCD_ConnectCallback(), and why is it configurable at all?

Reply:
`HAL_PCD_ConnectCallback` is called once, assuming VBUS sensing is enabled and the voltage is present (otherwise if there is no voltage it ping-pongs from `ConnectCallback` to `DisconnectCallback` until the voltage is detected).

The rationale behind this request is explained in #53236, I recommend to take a look at the application note linked in there.

Reply:
> The rationale behind this request is explained in #53236, I recommend to take a look at the application note linked in there.

It is not explained, it just copy&paste few world from AN4879, please justify here and in the commit message. Looks like it always depends on PA9, that should also be mentioned.  Also I think yes it should be DT property.

Reply:
I'll try to elaborate further, although it's unlikely I will be able to make a better job than ST's description in the AN.

VBUS sensing is mandatory for self-powered devices according to the USB specification, that means that if the MCU is using an external power input the USB driver will operate correctly only if the voltage sensing is enabled. The pin designated for this task, will always be PA9.

As an example, a host connected via USB to a self-powered Blackpill running the `cdc_acm` sample produces the dmesg output below only if the mandatory voltage is present in the sensing pin (PA9).

```
[  +3.683204] usb 1-4: new full-speed USB device number 44 using xhci_hcd
[  +0.150155] usb 1-4: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.02
[  +0.000014] usb 1-4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  +0.000006] usb 1-4: Product: Zephyr CDC ACM sample
[  +0.000006] usb 1-4: Manufacturer: ZEPHYR
[  +0.000005] usb 1-4: SerialNumber: 4E56500600590059
[  +0.003168] cdc_acm 1-4:1.0: ttyACM2: USB ACM device
```
Conversely, removing the voltage yields in a disconnection
```
[  +1.001391] usb 1-4: USB disconnect, device number 44
```
Please let me know if the clarification is still insufficient.

Would you mind elaborating on the reason why you think it should be a DT property and how that would compare to this proposal please?

Thanks

Reply:
Happy new year :smiley:

I have reworked the suggestion to make the Vbus detection selectable based on the DT instead of Kconfig which I hope lines up better with the request.

While looking into it, I did a search in the modules directory and found out that higher end ST MCUs (F/H7) offer HS and FS interfaces. In those cases, the Vbus detection is segregated into separate pins, namely PA9 & PB13. I have tried to capture both, albeit the driver implementation seems exclusive (see snippet at the end) so I am unsure whether having both enabled simultaneously would actually be practical. Also my hardware tests, though successful, are limited to the aforementioned MCU that comes with a single OTG FS peripheral (PA9).
The preprocessor output has been inspected in all the cases, based on selected drivers and pins, to ensure that the expanded macros are sensical at least.

```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otgfs) && DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
#error "Only one interface should be enabled at a time, OTG FS or OTG HS"
#endif
```